### ISSUE TITLE
Error loading module tds due to 'threads.sharedserialize'

### ISSUE THREAD

**Issue by andreykurenkov** [facebookresearch/deepmask#53]
Hi, 

I have installed DeepMask and have been able to run the pre-trained version on images, but cannot train it by running th train.lua . I get the following error:

andrey@regalalgorithm:~/Documents/Stanford/Project/deepmask$ th train.lua -nthreads 1
-- ignore option dm	
-- ignore option datadir	
nthreads	1	2	
-- ignore option rundir	
-- ignore option reload	
-- ignore option gpu	
| running in directory /home/andrey/Documents/Stanford/Project/deepmask/exps/deepmask/exp,nthreads=1	
| number of paramaters trunk: 15198016	
| number of paramaters mask branch: 1608768	
| number of paramaters score branch: 526337	
| number of paramaters total: 17333121	
/home/andrey/Documents/Libs/torch/install/bin/lua: ...Libs/torch/install/share/lua/5.1/threads/threads.lua:183: [thread 1 callback] .../Documents/Stanford/Project/deepmask/DataSampler.lua:12: loop or previous error loading module 'tds'
stack traceback:
	[C]: in function 'require'
	.../Documents/Stanford/Project/deepmask/DataSampler.lua:12: in main chunk
	[C]: in function 'dofile'
	...ents/Libs/torch/install/share/lua/5.1/paths/init.lua:84: in function 'dofile'
	...y/Documents/Stanford/Project/deepmask/DataLoader.lua:35: in function <...y/Documents/Stanford/Project/deepmask/DataLoader.lua:30>
	(tail call): ?
	[C]: in function 'xpcall'
	...Libs/torch/install/share/lua/5.1/threads/threads.lua:234: in function 'callback'
	...s/Libs/torch/install/share/lua/5.1/threads/queue.lua:65: in function <...s/Libs/torch/install/share/lua/5.1/threads/queue.lua:41>
	[C]: in function 'pcall'
	...s/Libs/torch/install/share/lua/5.1/threads/queue.lua:40: in function 'dojob'
	[string "  local Queue = require 'threads.queue'..."]:13: in main chunk
stack traceback:
	[C]: in function 'error'
	...Libs/torch/install/share/lua/5.1/threads/threads.lua:183: in function 'dojob'
	...Libs/torch/install/share/lua/5.1/threads/threads.lua:264: in function 'synchronize'
	...Libs/torch/install/share/lua/5.1/threads/threads.lua:142: in function 'specific'
	...Libs/torch/install/share/lua/5.1/threads/threads.lua:125: in function <...Libs/torch/install/share/lua/5.1/threads/threads.lua:36>
	(tail call): ?
	...y/Documents/Stanford/Project/deepmask/DataLoader.lua:40: in function '__init'
	...ents/Libs/torch/install/share/lua/5.1/torch/init.lua:91: in function <...ents/Libs/torch/install/share/lua/5.1/torch/init.lua:87>
	[C]: in function 'DataLoader'
	...y/Documents/Stanford/Project/deepmask/DataLoader.lua:21: in function 'create'
	train.lua:100: in main chunk
	[C]: in function 'dofile'
	.../torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: ?

At first it looked very similar to #41 , but I have tds installed and can import it with no problem elsewhere. After digging a bit I realized it may have to do with the line Threads.serialization('threads.sharedserialize') , and indeed removing it causes the problem to go away, but then I get another problem 

/home/andrey/Documents/Libs/torch/install/bin/lua: ...Libs/torch/install/share/lua/5.1/threads/threads.lua:183: [thread 1 callback] ...uments/Libs/torch/install/share/lua/5.1/tds/hash.lua:150: bad argument #1 to 'pairs' (table expected, got userdata)
stack traceback:
	[C]: in function 'pairs'
	...uments/Libs/torch/install/share/lua/5.1/tds/hash.lua:150: in function <...uments/Libs/torch/install/share/lua/5.1/tds/hash.lua:136>
	[C]: ?
	[C]: ?
	[C]: in function 'print'
	.../Documents/Stanford/Project/deepmask/DataSampler.lua:169: in function 'scoreSampling'
	.../Documents/Stanford/Project/deepmask/DataSampler.lua:113: in function 'get'
	...y/Documents/Stanford/Project/deepmask/DataLoader.lua:69: in function <...y/Documents/Stanford/Project/deepmask/DataLoader.lua:63>
	(tail call): ?
	[C]: in function 'xpcall'
	...Libs/torch/install/share/lua/5.1/threads/threads.lua:234: in function 'callback'
	...s/Libs/torch/install/share/lua/5.1/threads/queue.lua:65: in function <...s/Libs/torch/install/share/lua/5.1/threads/queue.lua:41>
	[C]: in function 'pcall'
	...s/Libs/torch/install/share/lua/5.1/threads/queue.lua:40: in function 'dojob'
	[string "  local Queue = require 'threads.queue'..."]:15: in main chunk
stack traceback:
	[C]: in function 'error'
	...Libs/torch/install/share/lua/5.1/threads/threads.lua:183: in function 'dojob'
	...y/Documents/Stanford/Project/deepmask/DataLoader.lua:94: in function '(for generator)'
	...uments/Stanford/Project/deepmask/TrainerDeepMask.lua:68: in function 'train'
	train.lua:116: in main chunk
	[C]: in function 'dofile'
	.../torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: ?

Have not been able to get a clue as to why this is happening... would appreciate help.

**Comment by andreykurenkov** [comment]
It appears that this error is due to the tds library not playing nicely with some versions of Lua (LuaJIT 2.1 and Lua 5.1 to be exact) - https://github.com/torch/tds/issues/6 . I switched Torch to LuaJIT 2.0 and was able to train fine.
